File: search_in_rotated.cpp
/*Time Complexity :O(log(n))
Space Complexity :O(1)
Did this code successfully run on Leetcode :Yes
Any problem you faced while coding this :No


Your code here along with comments explaining your approach
1) we find the middle element using binary search. Then determine if left half of right half is sorted.
2)If left half is sorted then check if the target lies within the left half range. If the target lies then high = mid-1
else low = mid +1
3) Do the same thing if the right half is sorted.
*/

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int low = 0;
        int high = nums.size()-1;
        while(low <= high){
            int mid = low + (high-low)/2;
            if(target == nums[mid])return mid;
            else if(nums[mid]>=nums[low])//left half is sorted
            {
                if(target < nums[mid] && target >=nums[low]){
                    high = mid-1;
                }
                else{
                    low = mid +1;
                }
            }
            else{
                if(target > nums[mid] && target <= nums[high]){
                    low = mid+1;
                }
                else{
                    high = mid-1;
                }
            }
        }
        return -1;

        
    }
};

File: SearchInRotated.py
class Solution:
    def search(self, nums, target) :

        def rotate_index(left, right):
            if nums[left] < nums[right]:
                return 0
            while left <= right:
                mid = (left + right) // 2
                if nums[mid] > nums[mid + 1]:
                    return mid + 1
                else:
                    if nums[mid] < nums[left]:
                        right = mid - 1
                    else:
                        left = mid + 1

        def search(left, right):

            while left <= right:

                mid = (left + right) // 2
                if nums[mid] == target:
                    return mid
                else:
                    if target < nums[mid]:
                        right = mid - 1
                    else:
                        left = mid + 1
            return -1

        length = len(nums)

        if length == 1:
            if nums[0] == target:
                return 0
            else:
                return -1

        r_index = rotate_index(0, length - 1)

        if nums[r_index] == 0:
            return search(0, length - 1)
        if r_index == 0:
            return search(0, length - 1)
        if target < nums[0]:
            return search(r_index, length - 1)
        return search(0, r_index)
# s= Solution()
# nums = [7,8,9,10,0,1,2,3,4,5,6]
# s.search(nums,9)